storage/src/data_access_layer/transfer.rs

```diff
@@ -0,0 +1,235 @@
+use std::collections::BTreeSet;
+
+use casper_types::{
+    account::AccountHash, execution::Effects, Digest, ProtocolVersion, PublicKey, RuntimeArgs,
+    TransferAddr, U512,
+};
+
+use crate::system::transfer::{TransferArgs, TransferError};
+
+/// Transfer details.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum TransferRequestArgs {
+    Raw(RuntimeArgs),
+    Explicit(TransferArgs),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum TransferConfig {
+    Administered {
+        administrative_accounts: BTreeSet<AccountHash>,
+        allow_unrestricted_transfer: bool,
+    },
+    Unadministered,
+}
+
+impl TransferConfig {
+    /// Returns a new instance.
+    pub fn new(
+        administrative_accounts: BTreeSet<AccountHash>,
+        allow_unrestricted_transfer: bool,
+    ) -> Self {
+        if administrative_accounts.is_empty() && allow_unrestricted_transfer {
+            TransferConfig::Unadministered
+        } else {
+            TransferConfig::Administered {
+                administrative_accounts,
+                allow_unrestricted_transfer,
+            }
+        }
+    }
+
+    /// Does account hash belong to an administrative account?
+    pub fn is_administrator(&self, account_hash: &AccountHash) -> bool {
+        match self {
+            TransferConfig::Administered {
+                administrative_accounts,
+                ..
+            } => administrative_accounts.contains(account_hash),
+            TransferConfig::Unadministered => false,
+        }
+    }
+
+    /// Administrative accounts, if any.
+    pub fn administrative_accounts(&self) -> BTreeSet<AccountHash> {
+        match self {
+            TransferConfig::Administered {
+                administrative_accounts,
+                ..
+            } => administrative_accounts.clone(),
+            TransferConfig::Unadministered => BTreeSet::default(),
+        }
+    }
+
+    /// Allow unrestricted transfers.
+    pub fn allow_unrestricted_transfers(&self) -> bool {
+        match self {
+            TransferConfig::Administered {
+                allow_unrestricted_transfer,
+                ..
+            } => *allow_unrestricted_transfer,
+            TransferConfig::Unadministered => true,
+        }
+    }
+
+    /// Restricted transfer should be enforced.
+    pub fn enforce_transfer_restrictions(&self, account_hash: &AccountHash) -> bool {
+        !self.allow_unrestricted_transfers() && !self.is_administrator(account_hash)
+    }
+}
+
+/// Request for motes transfer.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct TransferRequest {
+    /// Config.
+    config: TransferConfig,
+    /// State root hash.
+    state_hash: Digest,
+    /// Block time represented as a unix timestamp.
+    block_time: u64,
+    /// Protocol version.
+    protocol_version: ProtocolVersion,
+    /// Public key of the proposer.
+    proposer: PublicKey,
+    /// Transaction hash.
+    transaction_hash: Digest,
```

Comment:
Any reason not to use `casper_types::TransactionHash` here?

Reply:
it will be ultimately...as the PR comments note there's some missing versioned types for the go-forward. once fraser lands his bit, a fix up pass will address it.